File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/ui/client/pages/OrderDetailsPage.java
package com.mycompany.myshop.testkit.driver.adapter.ui.client.pages;

import com.mycompany.myshop.testkit.common.Converter;
import com.mycompany.myshop.testkit.driver.port.dtos.OrderStatus;
import com.mycompany.myshop.testkit.driver.adapter.shared.client.playwright.PageClient;

import java.math.BigDecimal;
import java.time.Instant;

public class OrderDetailsPage extends BasePage {
    private static final String ORDER_NUMBER_OUTPUT_SELECTOR = "[aria-label='Display Order Number']";
    private static final String ORDER_TIMESTAMP_OUTPUT_SELECTOR = "[aria-label='Display Order Timestamp']";
    private static final String SKU_OUTPUT_SELECTOR = "[aria-label='Display SKU']";
    private static final String COUNTRY_OUTPUT_SELECTOR = "[aria-label='Display Country']";
    private static final String QUANTITY_OUTPUT_SELECTOR = "[aria-label='Display Quantity']";
    private static final String UNIT_PRICE_OUTPUT_SELECTOR = "[aria-label='Display Unit Price']";
    private static final String BASE_PRICE_OUTPUT_SELECTOR = "[aria-label='Display Base Price']";
    private static final String SUBTOTAL_PRICE_OUTPUT_SELECTOR = "[aria-label='Display Subtotal Price']";
    private static final String DISCOUNT_RATE_OUTPUT_SELECTOR = "[aria-label='Display Discount Rate']";
    private static final String DISCOUNT_AMOUNT_OUTPUT_SELECTOR = "[aria-label='Display Discount Amount']";
    private static final String TAX_RATE_OUTPUT_SELECTOR = "[aria-label='Display Tax Rate']";
    private static final String TAX_AMOUNT_OUTPUT_SELECTOR = "[aria-label='Display Tax Amount']";
    private static final String TOTAL_PRICE_OUTPUT_SELECTOR = "[aria-label='Display Total Price']";
    private static final String STATUS_OUTPUT_SELECTOR = "[aria-label='Display Status']";
    private static final String APPLIED_COUPON_OUTPUT_SELECTOR = "[aria-label='Display Applied Coupon']";
    private static final String CANCEL_ORDER_SELECTOR = "[aria-label='Cancel Order']";
    private static final String DELIVER_ORDER_SELECTOR = "[aria-label='Deliver Order']";
    private static final String TEXT_NONE = "None";
    private static final String DOLLAR_SYMBOL = "$";
    private static final String PERCENT_SYMBOL = "%";

    public OrderDetailsPage(PageClient pageClient) {
        super(pageClient);
    }

    public boolean isLoadedSuccessfully() {
        return pageClient.isVisible(ORDER_NUMBER_OUTPUT_SELECTOR);
    }

    public String getOrderNumber() {
        return pageClient.readTextContent(ORDER_NUMBER_OUTPUT_SELECTOR);
    }

    public Instant getOrderTimestamp() {
        var textContent = pageClient.readTextContent(ORDER_TIMESTAMP_OUTPUT_SELECTOR);
        return Converter.toInstant(textContent);
    }

    public String getSku() {
        return pageClient.readTextContent(SKU_OUTPUT_SELECTOR);
    }

    public String getCountry() {
        return pageClient.readTextContent(COUNTRY_OUTPUT_SELECTOR);
    }

    public int getQuantity() {
        var textContent = pageClient.readTextContent(QUANTITY_OUTPUT_SELECTOR);
        return Integer.parseInt(textContent);
    }

    public BigDecimal getUnitPrice() {
        return readTextMoney(UNIT_PRICE_OUTPUT_SELECTOR);
    }

    public BigDecimal getBasePrice() {
        return readTextMoney(BASE_PRICE_OUTPUT_SELECTOR);
    }

    public BigDecimal getDiscountRate() {
        return readTextPercentage(DISCOUNT_RATE_OUTPUT_SELECTOR);
    }

    public BigDecimal getDiscountAmount() {
        return readTextMoney(DISCOUNT_AMOUNT_OUTPUT_SELECTOR);
    }

    public BigDecimal getSubtotalPrice() {
        return readTextMoney(SUBTOTAL_PRICE_OUTPUT_SELECTOR);
    }

    public BigDecimal getTaxRate() {
        return readTextPercentage(TAX_RATE_OUTPUT_SELECTOR);
    }

    public BigDecimal getTaxAmount() {
        return readTextMoney(TAX_AMOUNT_OUTPUT_SELECTOR);
    }

    public BigDecimal getTotalPrice() {
        return readTextMoney(TOTAL_PRICE_OUTPUT_SELECTOR);
    }

    public OrderStatus getStatus() {
        var status = pageClient.readTextContent(STATUS_OUTPUT_SELECTOR);
        return OrderStatus.valueOf(status);
    }

    public String getAppliedCoupon() {
        var coupon = pageClient.readTextContent(APPLIED_COUPON_OUTPUT_SELECTOR);
        return TEXT_NONE.equals(coupon) ? null : coupon;
    }

    public void clickCancelOrder() {
        pageClient.click(CANCEL_ORDER_SELECTOR);
    }

    public void clickDeliverOrder() {
        pageClient.click(DELIVER_ORDER_SELECTOR);
    }

    public boolean isCancelButtonHidden() {
        return pageClient.isHidden(CANCEL_ORDER_SELECTOR);
    }

    private BigDecimal readTextMoney(String selector) {
        var textContent = pageClient.readTextContent(selector);
        var cleaned = textContent.replace(DOLLAR_SYMBOL, "").trim();
        return new BigDecimal(cleaned);
    }

    private BigDecimal readTextPercentage(String selector) {
        var textContent = pageClient.readTextContent(selector);
        var cleaned = textContent.replace(PERCENT_SYMBOL, "").trim();
        var value = new BigDecimal(cleaned);
        return value.divide(BigDecimal.valueOf(100));
    }
}

File: system-test/typescript/src/testkit/driver/adapter/ui/client/pages/OrderDetailsPage.ts
import { BasePage, PAGE_TIMEOUT_MS } from './BasePage.js';

export class OrderDetailsPage extends BasePage {
  async waitForLoad(): Promise<void> {
    await this.page
      .locator("[aria-label='Display Order Number']")
      .waitFor({ state: 'visible', timeout: PAGE_TIMEOUT_MS });
  }

  async clickCancelOrder(): Promise<void> {
    const cancelButton = this.page.locator('[aria-label="Cancel Order"]');
    await cancelButton.waitFor({ state: 'visible', timeout: PAGE_TIMEOUT_MS });
    await cancelButton.click({ timeout: PAGE_TIMEOUT_MS });
  }

  async clickDeliverOrder(): Promise<void> {
    const deliverButton = this.page.locator('[aria-label="Deliver Order"]');
    await deliverButton.waitFor({ state: 'visible', timeout: PAGE_TIMEOUT_MS });
    await deliverButton.click({ timeout: PAGE_TIMEOUT_MS });
  }

  async getOrderNumber(): Promise<string> {
    return (
      (await this.page.locator("[aria-label='Display Order Number']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || ''
    );
  }

  async getOrderTimestamp(): Promise<string> {
    return (
      (await this.page.locator("[aria-label='Display Order Timestamp']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() ||
      ''
    );
  }

  async getSku(): Promise<string> {
    return (
      (await this.page.locator("[aria-label='Display SKU']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || ''
    );
  }

  async getQuantity(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Quantity']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return parseInt(text, 10);
  }

  async getUnitPrice(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Unit Price']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return Number.parseFloat(text.replace('$', ''));
  }

  async getBasePrice(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Base Price']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return Number.parseFloat(text.replace('$', ''));
  }

  async getDiscountRate(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Discount Rate']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return Number.parseFloat(text.replace('%', '')) / 100;
  }

  async getDiscountAmount(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Discount Amount']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return Number.parseFloat(text.replace('$', ''));
  }

  async getSubtotalPrice(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Subtotal Price']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return Number.parseFloat(text.replace('$', ''));
  }

  async getTaxRate(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Tax Rate']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return Number.parseFloat(text.replace('%', '')) / 100;
  }

  async getTaxAmount(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Tax Amount']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return Number.parseFloat(text.replace('$', ''));
  }

  async getTotalPrice(): Promise<number> {
    const text =
      (await this.page.locator("[aria-label='Display Total Price']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '0';
    return Number.parseFloat(text.replace('$', ''));
  }

  async getCountry(): Promise<string> {
    return (
      (await this.page.locator("[aria-label='Display Country']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || ''
    );
  }

  async getAppliedCouponCode(): Promise<string | null> {
    const text = (await this.page.locator("[aria-label='Display Applied Coupon']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || '';
    return text === 'None' ? null : text;
  }

  async getStatus(): Promise<string> {
    return (
      (await this.page.locator("[aria-label='Display Status']").textContent({ timeout: PAGE_TIMEOUT_MS }))?.trim() || ''
    );
  }
}
